include/RED4ext/Callback.hpp

```diff
@@ -0,0 +1,382 @@
+    {
+        handler = nullptr;
+    }
+
+    R InvokeTarget(Args&&... aArgs) const
+    {
+        return (handler->invoke)(buffer, std::forward<Args>(aArgs)...);
+    }
+
+    void CopyTargetFrom(void* aTarget)
+    {
+        if (handler)
+        {
+            handler->copy(buffer, aTarget);
+        }
+    }
+
+    void MoveTargetFrom(void* aTarget)
+    {
+        if (handler)
+        {
+            handler->move(buffer, aTarget);
+        }
+    }
+
+    void DestroyTarget()
+    {
+        if (handler)
+        {
+            handler->destruct(buffer);
+        }
+    }
+};
+RED4EXT_ASSERT_SIZE(Callback<void (*)()>, DefaultFixedCallbackBufferSize + 0x8);
+RED4EXT_ASSERT_OFFSET(Callback<void (*)()>, handler, DefaultFixedCallbackBufferSize);
+
+template<typename T, size_t InlineSize = DefaultFlexCallbackBufferSize>
+class FlexCallback;
+
+template<typename R, typename... Args, size_t InlineSize>
+class FlexCallback<R (*)(Args...), InlineSize>
+{
+public:
+    using HandlerPtr = CallbackHandler<R, Args...>*;
+    using AllocatorType = Memory::DefaultAllocator;
+
+    static_assert(InlineSize > sizeof(void*), "Buffer size can't be less than pointer size");
+
+    FlexCallback(R (*aFunc)(Args...))
+    {
+        using TargetType = Detail::UnboundFunctionTarget<R, Args...>;
+
+        TargetType target{aFunc};
+        InitializeBuffer(sizeof(TargetType));
+        InitializeHandler(&target);
+    }
+
+    template<typename C>
+    FlexCallback(C* aContext, R (C::*aFunc)(Args...))
+    {
+        using TargetType = Detail::MemberFunctionTarget<C, R, Args...>;
+
+        TargetType target{aContext, aFunc};
+        InitializeBuffer(sizeof(TargetType));
+        InitializeHandler(&target);
+    }
+
+    template<typename L>
+    requires Detail::IsClosure<L, R, Args...>
+    FlexCallback(L&& aClosure)
+    {
+        using TargetType = Detail::ClosureTarget<L, R, Args...>;
+
+        InitializeBuffer(sizeof(TargetType));
+        InitializeHandler(reinterpret_cast<TargetType*>(&aClosure));
+    }
+
+    FlexCallback(const FlexCallback& aOther)
+    {
+        InitializeBuffer(aOther.GetSize());
+        CopyHandlerFrom(aOther.handler);
+        CopyTargetFrom(aOther.GetBuffer());
+    }
+
+    FlexCallback(FlexCallback&& aOther) noexcept
+    {
+        InitializeBuffer(aOther.GetSize());
+        MoveHandlerFrom(aOther.handler);
+        MoveTargetFrom(aOther.GetBuffer());
+    }
+
+    ~FlexCallback()
+    {
+        DestroyTarget();
+        ResetBuffer();
+        ResetHandler();
+    }
+
+    R operator()(Args&&... aArgs) const
+    {
+        return InvokeTarget(std::forward<Args>(aArgs)...);
+    }
+
+    uint8_t buffer[InlineSize];
+    HandlerPtr handler;
+    Memory::IAllocator* allocator;
+    uint32_t extendedSize;
+
+protected:
+    static constexpr uint32_t ExtendedFlag = 0x80000000;
+    static constexpr uint32_t SizeMask = ~ExtendedFlag;
+
+    template<class TargetType>
+    void InitializeHandler(TargetType* aSrc) noexcept
```

Comment:
`FlexCallback` shouldn't be subtype of `Callback`. And common base is pretty small to introduce `CallbackBase` imo.

Reply:
I would do a base class just to make sure in the future we won't do some changes in one and not the other. But it is up to you if you want to do it.

Reply:
Good point. But I'd leave it as is for sake of clearer implementation. 
Most functions suitable for the base class need to be rewritten to take one more parameter for a pointer to a buffer. `Callback` always uses `buffer, and `FlexCallback` always uses `GetBuffer()` (and it's not virtual, it can't be part of the base class, it must be passed to every base function that accesses the buffer). I'm afraid this'll make it harder to understand the `Callback`.

Reply:
Good point. Then let's keep it as it is.